File: Sorts/python/merge_sort.py
def merge_sort(list):
    # check if the len is greater then 1
    if len(list) > 1:
        # get the middle point
        middle = len(list) / 2
        # split the list half
        leftHalf = list[:middle]
        rightHalf = list[middle:]
        # call merge sort again on the two half
        merge_sort(leftHalf)
        merge_sort(rightHalf)

        # merge the half together
        merge(list, leftHalf, rightHalf)


def merge(list, leftHalf, rightHalf):
    # now merge them together
    i = 0
    j = 0
    k = 0
    while i < len(leftHalf) and j < len(rightHalf):
        if(leftHalf[i] < rightHalf[j]):
            list[k] = leftHalf[i]
            i += 1
        else:
            list[k] = rightHalf[j]
            j += 1
        k += 1

    # if the leftHalt still have values
    while i < len(leftHalf):
        list[k] = leftHalf[i]
        i += 1
        k += 1
    
    # if the right still have values
    while j < len(rightHalf):
        list[k] = rightHalf[j]
        j += 1
        k += 1


myList = [5, 15, 2, 20, 11, 16, 4, 14, 6, 3, 8]
print("Before: ", myList)
merge_sort(myList)
print("After: ", myList)

File: Sorts/js/merge_sort.js
function merge_sort(arr) {
  if (arr.length > 1) {
    let mid = Math.floor(arr.length / 2);
    // split the array into half
    let leftHalf = arr.slice(0, mid);
    let rightHalf = arr.slice(mid);

    // call merge sort again
    merge_sort(leftHalf);
    merge_sort(rightHalf);

    // merge these two sorted arrays
    merge(arr, leftHalf, rightHalf);
  }
}

function merge(arr, leftHalf, rightHalf) {
  let i = (j = k = 0); // same as i = 0; j = 0; k = 0;
  while (i < leftHalf.length && j < rightHalf.length) {
    if (leftHalf[i] < rightHalf[j]) {
      arr[k] = leftHalf[i];
      i++;
    } else {
      arr[k] = rightHalf[j];
      j++;
    }
    k++;
  }

  // check if leftHalf still has values
  while (i < leftHalf.length) {
    arr[k] = leftHalf[i];
    i++;
    k++;
  }
  // check if right haft still has values
  while (j < rightHalf.length) {
    arr[k] = rightHalf[j];
    j++;
    k++;
  }
}

arr = [5, 15, 2, 20, 11, 16, 4, 14, 6, 3, 8];
console.log("before: ", arr);
merge_sort(arr);
console.log("after: ", arr);
